test/adapter/json_api/key_case_test.rb

```diff
@@ -0,0 +1,323 @@
+          adapter = ActiveModelSerializers::Adapter::JsonApi.new(serializer)
+          result = adapter.serializable_hash(@options)
+          assert_equal({
+            Data: {
+              Id: '1337',
+              Type: 'posts',
+              Attributes: {
+                Title: 'Title 1',
+                Body: 'Body 1',
+                PublishAt: @publish_at
+              },
+              Relationships: {
+                Author: {
+                  Data: { Id: '1', Type: 'authors' }
+                },
+                Comments: {
+                  Data: [
+                    { Id: '7', Type: 'comments' },
+                    { Id: '12', Type: 'comments' }
+                ] }
+              },
+              Links: {
+                Self: 'http://example.com/posts/1337',
+                PostAuthors: 'http://example.com/posts/1337/authors',
+                SubscriberComments: 'http://example.com/posts/1337/comments'
+              },
+              Meta: { Rating: 5, FavoriteCount: 10 }
+            }
+          }, result)
+        end
+
+        def test_success_key_case_camel_lower
+          mock_request(:camel_lower)
+          serializer = PostSerializer.new(@post)
+          adapter = ActiveModelSerializers::Adapter::JsonApi.new(serializer)
+          result = adapter.serializable_hash(@options)
+          assert_equal({
+            data: {
+              id: '1337',
+              type: 'posts',
+              attributes: {
+                title: 'Title 1',
+                body: 'Body 1',
+                publishAt: @publish_at
+              },
+              relationships: {
+                author: {
+                  data: { id: '1', type: 'authors' }
+                },
+                comments: {
+                  data: [
+                    { id: '7', type: 'comments' },
+                    { id: '12', type: 'comments' }
+                ] }
+              },
+              links: {
+                self: 'http://example.com/posts/1337',
+                postAuthors: 'http://example.com/posts/1337/authors',
+                subscriberComments: 'http://example.com/posts/1337/comments'
+              },
+              meta: { rating: 5, favoriteCount: 10 }
+            }
+          }, result)
+        end
+
+        def test_error_document_key_case_dashed
+          mock_request(:dashed)
+
+          resource = ModelWithErrors.new
+          resource.errors.add(:published_at, 'must be in the future')
+          resource.errors.add(:title, 'must be longer')
+
+          serializer = ActiveModel::Serializer::ErrorSerializer.new(resource)
+          adapter = ActiveModelSerializers::Adapter::JsonApi.new(serializer)
+          result = adapter.serializable_hash(@options)
+
+          expected_errors_object =
+            { :errors =>
+              [
+                { :source => { :pointer => '/data/attributes/published_at' }, :detail => 'must be in the future' },
```

Comment:
This one is tricky because we're only modifying keys at the end of the document building and this is in a value. I figured we could look more into it when the error document gets links and meta support.